File: apps/shared/src/commonTest/kotlin/dev/supermux/net/TerminalProtocolTest.kt
package dev.supermux.net

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertTrue

/**
 * Revision-2 wire fixtures. THE SAME STRINGS ARE ENCODED/DECODED BY THE BROKER:
 * `tests/terminal-protocol.test.ts` holds them verbatim and reads THIS file,
 * failing when a fixture drifts out of one side. A protocol both ends only
 * believe they share is how revision 1 ended up with a client that matched
 * frames by substring.
 */
private object Fixtures {
    // server → client
    const val READY = """{"type":"ready","version":2,"epoch":"c-1","replyOwner":false,"ownerGeneration":0}"""
    const val READY_UNKNOWN_VERSION = """{"type":"ready","version":3,"epoch":"c-1","replyOwner":false,"ownerGeneration":0}"""
    const val RESET_A = """{"type":"reset","epoch":"e-a"}"""
    const val REPLAY_START_A = """{"type":"replay-start","epoch":"e-a"}"""
    const val REPLAY_END_A = """{"type":"replay-end","epoch":"e-a"}"""
    const val RESET_B = """{"type":"reset","epoch":"e-b"}"""
    const val REPLAY_START_B = """{"type":"replay-start","epoch":"e-b"}"""
    const val REPLAY_END_B = """{"type":"replay-end","epoch":"e-b"}"""
    const val OWNER_ON = """{"type":"owner","epoch":"e-a","enabled":true,"ownerGeneration":1}"""
    const val OWNER_STALE_EPOCH = """{"type":"owner","epoch":"e-gone","enabled":true,"ownerGeneration":9}"""
    const val REPLAY_START_STALE_EPOCH = """{"type":"replay-start","epoch":"e-gone"}"""
    const val EXIT_NON_ZERO = """{"type":"exit","known":true,"code":3,"signal":null}"""
    const val EXIT_UNKNOWN = """{"type":"exit","known":false,"code":null,"signal":null}"""
    const val EXIT_SIGNALLED = """{"type":"exit","known":true,"code":null,"signal":9}"""
    const val FAILURE_RECOVERABLE_SAYING_EXIT =
        """{"type":"failure","code":"backend-unavailable","recoverable":true,"message":"zmx helper exited before the attach completed"}"""
    const val FAILURE_FATAL = """{"type":"failure","code":"target-not-found","recoverable":false,"message":"no such terminal"}"""
    const val FAILURE_UNSUPPORTED_REVISION =
        """{"type":"failure","code":"protocol-unsupported","recoverable":false,"message":"terminal protocol \"1\" is not supported; this broker speaks revision 2"}"""
    const val MALFORMED = """{"type":"reset","epoch":"""
    const val UNKNOWN_TYPE = """{"type":"bell","epoch":"e-a"}"""

    // client → server
    const val RESIZE = """{"type":"resize","cols":120,"rows":40}"""
    const val FOCUS = """{"type":"focus","focused":true,"cols":120,"rows":40}"""
    const val BLUR = """{"type":"focus","focused":false}"""
    const val REPLY = """{"type":"reply","epoch":"e-a","ownerGeneration":1,"data":"G1syNDsxUg=="}"""
    const val CLOSE = """{"type":"close"}"""
    const val RESIZE_NOT_A_NUMBER = """{"type":"resize","cols":"120","rows":40}"""
}

/** The bytes of "ok→\n" — the arrow is 3 bytes and the fixture splits it. */
private val UTF8_CHUNK_A = byteArrayOf(0x6f, 0x6b, 0xe2.toByte())
private val UTF8_CHUNK_B = byteArrayOf(0x86.toByte(), 0x92.toByte(), 0x0a)

private fun TerminalDecode.event(): TerminalEvent {
    assertIs<TerminalDecode.Deliver>(this)
    return event
}

class TerminalProtocolTest {

    @Test fun a_fresh_connection_is_ready_reset_replay_output_replay_end_then_live_output() {
        val d = TerminalEventDecoder()
        assertEquals(
            TerminalEvent.Ready(2, "c-1", replyOwner = false, ownerGeneration = 0),
            d.onText(Fixtures.READY).event(),
        )
        assertEquals(TerminalEvent.Reset("e-a"), d.onText(Fixtures.RESET_A).event())
        assertEquals(TerminalEvent.ReplayStart("e-a"), d.onText(Fixtures.REPLAY_START_A).event())
        // Chunked UTF-8: the client NEVER decodes text, it forwards bytes. A
        // multi-byte character split across two frames must arrive split.
        assertEquals(TerminalEvent.Output(UTF8_CHUNK_A), d.onBinary(UTF8_CHUNK_A).event())
        assertEquals(TerminalEvent.Output(UTF8_CHUNK_B), d.onBinary(UTF8_CHUNK_B).event())
        assertEquals(TerminalEvent.ReplayEnd("e-a"), d.onText(Fixtures.REPLAY_END_A).event())
        assertTrue(d.replayClosed)
        assertEquals(TerminalEvent.Output(byteArrayOf(0x24)), d.onBinary(byteArrayOf(0x24)).event())
        assertEquals("ok→\n", (UTF8_CHUNK_A + UTF8_CHUNK_B).decodeToString())
    }

    @Test fun a_replay_may_contain_no_output_at_all() {
        val d = TerminalEventDecoder()
        d.onText(Fixtures.READY)
        d.onText(Fixtures.RESET_A)
        d.onText(Fixtures.REPLAY_START_A)
        assertEquals(TerminalEvent.ReplayEnd("e-a"), d.onText(Fixtures.REPLAY_END_A).event())
        assertTrue(d.replayClosed)
    }

    @Test fun output_before_ready_is_refused() {
        val d = TerminalEventDecoder()
        assertIs<TerminalDecode.Ignore>(d.onBinary(byteArrayOf(1, 2, 3)))
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.RESET_A))
        d.onText(Fixtures.READY)
        assertEquals(TerminalEvent.Output(byteArrayOf(1, 2, 3)), d.onBinary(byteArrayOf(1, 2, 3)).event())
    }

    @Test fun a_failure_that_mentions_exiting_is_not_an_exit() {
        val d = TerminalEventDecoder()
        d.onText(Fixtures.READY)
        val event = d.onText(Fixtures.FAILURE_RECOVERABLE_SAYING_EXIT).event()
        assertEquals(
            TerminalEvent.Failure("backend-unavailable", recoverable = true, message = "zmx helper exited before the attach completed"),
            event,
        )
        assertTrue(event !is TerminalEvent.Exit)
        // ...and the non-recoverable one is still a failure, not an exit.
        assertEquals(
            TerminalEvent.Failure("target-not-found", recoverable = false, message = "no such terminal"),
            d.onText(Fixtures.FAILURE_FATAL).event(),
        )
        assertEquals(
            "protocol-unsupported",
            (d.onText(Fixtures.FAILURE_UNSUPPORTED_REVISION).event() as TerminalEvent.Failure).code,
        )
    }

    @Test fun malformed_and_unknown_control_frames_are_dropped_and_the_stream_survives() {
        val d = TerminalEventDecoder()
        d.onText(Fixtures.READY)
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.MALFORMED))
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.UNKNOWN_TYPE))
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.RESIZE_NOT_A_NUMBER))
        assertIs<TerminalDecode.Ignore>(d.onText("not json at all"))
        // The connection is intact: the next real frame still decodes.
        assertEquals(TerminalEvent.Reset("e-a"), d.onText(Fixtures.RESET_A).event())
    }

    @Test fun an_unknown_protocol_version_is_fatal_and_not_retried() {
        val d = TerminalEventDecoder()
        val decoded = d.onText(Fixtures.READY_UNKNOWN_VERSION)
        assertIs<TerminalDecode.Fatal>(decoded)
        assertEquals("protocol-unsupported", decoded.code)
        assertTrue(!d.ready)
    }

    @Test fun a_reset_mid_output_opens_a_new_epoch_and_voids_the_old_one() {
        val d = TerminalEventDecoder()
        d.onText(Fixtures.READY)
        d.onText(Fixtures.RESET_A)
        d.onText(Fixtures.REPLAY_START_A)
        d.onText(Fixtures.REPLAY_END_A)
        d.onBinary(UTF8_CHUNK_A)
        // The backend re-synchronised mid-stream.
        assertEquals(TerminalEvent.Reset("e-b"), d.onText(Fixtures.RESET_B).event())
        assertEquals("e-b", d.epoch)
        assertTrue(!d.replayClosed, "a new epoch re-opens the replay boundary")
        // Frames still addressed to the epoch that was just replaced are
        // history: acting on them would draw the old screen over the new one.
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.REPLAY_START_A))
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.REPLAY_START_STALE_EPOCH))
        assertEquals(TerminalEvent.ReplayStart("e-b"), d.onText(Fixtures.REPLAY_START_B).event())
        assertEquals(TerminalEvent.ReplayEnd("e-b"), d.onText(Fixtures.REPLAY_END_B).event())
    }

    @Test fun exit_carries_code_signal_and_whether_anybody_reaped_it() {
        fun exitOf(fixture: String): TerminalEvent.Exit {
            val d = TerminalEventDecoder()
            d.onText(Fixtures.READY)
            return d.onText(fixture).event() as TerminalEvent.Exit
        }
        assertEquals(TerminalEvent.Exit(code = 3, signal = null, known = true), exitOf(Fixtures.EXIT_NON_ZERO))
        // Nobody reaped a status: 0 would be a clean exit nobody observed.
        assertEquals(TerminalEvent.Exit(code = null, signal = null, known = false), exitOf(Fixtures.EXIT_UNKNOWN))
        assertEquals(TerminalEvent.Exit(code = null, signal = 9, known = true), exitOf(Fixtures.EXIT_SIGNALLED))
    }

    @Test fun owner_frames_are_accepted_on_the_live_epoch_and_a_rising_generation() {
        val d = TerminalEventDecoder()
        d.onText(Fixtures.READY)
        d.onText(Fixtures.RESET_A)
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.OWNER_STALE_EPOCH))
        assertTrue(!d.replyOwner)
        assertEquals(TerminalEvent.Owner("e-a", enabled = true, ownerGeneration = 1), d.onText(Fixtures.OWNER_ON).event())
        assertTrue(d.replyOwner)
        assertEquals(1L, d.ownerGeneration)
        // A generation we already passed is a reordered frame, not a lease.
        assertIs<TerminalDecode.Ignore>(d.onText(Fixtures.OWNER_ON))
        assertTrue(d.replyOwner)
    }

    @Test fun viewer_commands_encode_to_the_bytes_the_broker_decodes() {
        assertEquals(Fixtures.RESIZE, encodeTerminalCommand(TerminalCommand.Resize(120, 40)))
        assertEquals(Fixtures.FOCUS, encodeTerminalCommand(TerminalCommand.Focus(true, 120, 40)))
        assertEquals(Fixtures.BLUR, encodeTerminalCommand(TerminalCommand.Focus(false)))
        assertEquals(Fixtures.CLOSE, encodeTerminalCommand(TerminalCommand.Close))
        assertEquals(
            Fixtures.REPLY,
            encodeTerminalCommand(
                TerminalCommand.Reply("e-a", 1, encodeReplyPayload("\u001b[24;1R".encodeToByteArray())),
            ),
        )
    }
}

File: tests/terminal-protocol.test.ts
// Revision-2 wire fixtures. THE SAME STRINGS ARE DECODED BY THE KOTLIN CLIENT:
// `apps/shared/src/commonTest/kotlin/dev/supermux/net/TerminalProtocolTest.kt`
// holds them verbatim, and the last test in this file reads that file and
// fails if a fixture drifts out of one side. A protocol both ends only
// *believe* they share is how revision 1 ended up with a client that matched
// frames by substring.

import { test, expect } from "bun:test"
import { readFileSync } from "fs"
import { join } from "path"
import {
  TERMINAL_PROTOCOL_VERSION,
  TerminalFrameLane,
  decodeClientControl,
  decodeReplyPayload,
  encodeServerControl,
  parseTerminalRevision,
  type TerminalFrameSink,
  type TerminalServerControl,
} from "../src/channels/web/terminal-protocol"

// ── server → client ────────────────────────────────────────────────────────
export const SERVER_FIXTURES = {
  ready: `{"type":"ready","version":2,"epoch":"c-1","replyOwner":false,"ownerGeneration":0}`,
  readyUnknownVersion: `{"type":"ready","version":3,"epoch":"c-1","replyOwner":false,"ownerGeneration":0}`,
  resetA: `{"type":"reset","epoch":"e-a"}`,
  replayStartA: `{"type":"replay-start","epoch":"e-a"}`,
  replayEndA: `{"type":"replay-end","epoch":"e-a"}`,
  resetB: `{"type":"reset","epoch":"e-b"}`,
  replayStartB: `{"type":"replay-start","epoch":"e-b"}`,
  replayEndB: `{"type":"replay-end","epoch":"e-b"}`,
  ownerOn: `{"type":"owner","epoch":"e-a","enabled":true,"ownerGeneration":1}`,
  ownerStaleEpoch: `{"type":"owner","epoch":"e-gone","enabled":true,"ownerGeneration":9}`,
  replayStartStaleEpoch: `{"type":"replay-start","epoch":"e-gone"}`,
  exitNonZero: `{"type":"exit","known":true,"code":3,"signal":null}`,
  exitUnknown: `{"type":"exit","known":false,"code":null,"signal":null}`,
  exitSignalled: `{"type":"exit","known":true,"code":null,"signal":9}`,
  // The word "exit" inside a MESSAGE. Revision 1's client matched
  // `contains("\"type\":\"exit\"")`, so text like this closed the tab on what
  // is really a reconnect.
  failureRecoverableSayingExit: `{"type":"failure","code":"backend-unavailable","recoverable":true,"message":"zmx helper exited before the attach completed"}`,
  failureFatal: `{"type":"failure","code":"target-not-found","recoverable":false,"message":"no such terminal"}`,
  failureUnsupportedRevision: `{"type":"failure","code":"protocol-unsupported","recoverable":false,"message":"terminal protocol \\"1\\" is not supported; this broker speaks revision 2"}`,
  malformed: `{"type":"reset","epoch":`,
  unknownType: `{"type":"bell","epoch":"e-a"}`,
} as const

// ── client → server ────────────────────────────────────────────────────────
export const CLIENT_FIXTURES = {
  resize: `{"type":"resize","cols":120,"rows":40}`,
  focus: `{"type":"focus","focused":true,"cols":120,"rows":40}`,
  blur: `{"type":"focus","focused":false}`,
  // DSR cursor-position answer (ESC [ 24 ; 1 R) for epoch e-a, generation 1.
  reply: `{"type":"reply","epoch":"e-a","ownerGeneration":1,"data":"G1syNDsxUg=="}`,
  close: `{"type":"close"}`,
  resizeNotANumber: `{"type":"resize","cols":"120","rows":40}`,
} as const

/** The bytes of "ok→" — the arrow is 3 bytes and the fixtures split it. */
const UTF8_CHUNK_A = Uint8Array.from([0x6f, 0x6b, 0xe2])
const UTF8_CHUNK_B = Uint8Array.from([0x86, 0x92, 0x0a])

function recorder() {
  const frames: Array<string | Uint8Array> = []
  const closes: Array<{ code: number; reason: string }> = []
  const sink: TerminalFrameSink = {
    text: (payload) => { frames.push(payload) },
    binary: (bytes) => { frames.push(bytes) },
    close: (code, reason) => { closes.push({ code, reason }) },
  }
  return { frames, closes, sink }
}

/** A sink whose binary writes block until the returned release() is called. */
function blockingRecorder() {
  const frames: Array<string | Uint8Array> = []
  const closes: Array<{ code: number; reason: string }> = []
  const pending: Array<() => void> = []
  const sink: TerminalFrameSink = {
    text: (payload) => { frames.push(payload) },
    binary: (bytes) => {
      frames.push(bytes)
      return new Promise<void>((resolve) => { pending.push(resolve) })
    },
    close: (code, reason) => { closes.push({ code, reason }) },
  }
  return { frames, closes, sink, releaseAll: () => { for (const p of pending.splice(0)) p() } }
}

test("revision negotiation: 2 is spoken, absent is legacy, anything else is refused by name", () => {
  expect(parseTerminalRevision("2")).toEqual({ ok: true, revision: 2 })
  expect(parseTerminalRevision(null)).toEqual({ ok: true, revision: 1 })
  const old = parseTerminalRevision("1")
  expect(old.ok).toBe(false)
  if (!old.ok) {
    expect(encodeServerControl({ type: "failure", code: "protocol-unsupported", recoverable: false, message: old.message }))
      .toBe(SERVER_FIXTURES.failureUnsupportedRevision)
  }
  expect(parseTerminalRevision("banana").ok).toBe(false)
  expect(TERMINAL_PROTOCOL_VERSION).toBe(2)
})

test("a fresh connection is ready → reset → replay-start → output* → replay-end → output*", async () => {
  const { frames, sink } = recorder()
  const lane = new TerminalFrameLane(sink, "c-1")
  await lane.ready()
  await lane.event({ type: "reset", epoch: "e-a" })
  await lane.event({ type: "replay-start", epoch: "e-a" })
  await lane.event({ type: "output", bytes: UTF8_CHUNK_A })
  await lane.event({ type: "output", bytes: UTF8_CHUNK_B })
  await lane.event({ type: "replay-end", epoch: "e-a" })
  await lane.event({ type: "output", bytes: Uint8Array.from([0x24]) })

  expect(frames[0]).toBe(SERVER_FIXTURES.ready)
  expect(frames[1]).toBe(SERVER_FIXTURES.resetA)
  expect(frames[2]).toBe(SERVER_FIXTURES.replayStartA)
  expect(frames[3]).toEqual(UTF8_CHUNK_A)
  expect(frames[4]).toEqual(UTF8_CHUNK_B)
  expect(frames[5]).toBe(SERVER_FIXTURES.replayEndA)
  expect(frames[6]).toEqual(Uint8Array.from([0x24]))
  // A replay with no output at all is legal: the boundary still closes.
  const empty = recorder()
  const lane2 = new TerminalFrameLane(empty.sink, "c-2")
  await lane2.ready()
  await lane2.event({ type: "reset", epoch: "e-a" })
  await lane2.event({ type: "replay-start", epoch: "e-a" })
  await lane2.event({ type: "replay-end", epoch: "e-a" })
  expect(empty.frames.filter((f) => typeof f !== "string")).toEqual([])
  expect(empty.frames[3]).toBe(SERVER_FIXTURES.replayEndA)
})

test("a reset queued behind a stalled output frame still lands after it", async () => {
  const { frames, sink, releaseAll } = blockingRecorder()
  const lane = new TerminalFrameLane(sink, "c-1")
  await lane.ready()
  void lane.event({ type: "output", bytes: UTF8_CHUNK_A })
  // Queued while the write above is still stalled on the socket's drain. A
  // second callback writing straight to the socket would overtake it here.
  const reset = lane.event({ type: "reset", epoch: "e-b" })
  await new Promise((r) => setTimeout(r, 0))
  expect(frames).toEqual([SERVER_FIXTURES.ready, UTF8_CHUNK_A])
  releaseAll()
  await reset
  expect(frames).toEqual([SERVER_FIXTURES.ready, UTF8_CHUNK_A, SERVER_FIXTURES.resetB])
  expect(lane.epoch).toBe("e-b")
})

test("exit carries known/code/signal and closes; failure never masquerades as one", async () => {
  const a = recorder()
  const lane = new TerminalFrameLane(a.sink, "c-1")
  await lane.event({ type: "exit", known: true, code: 3, signal: null })
  expect(a.frames).toEqual([SERVER_FIXTURES.exitNonZero])
  expect(a.closes).toEqual([{ code: 1000, reason: "terminal exited" }])
  // A second terminal event after the connection is finished is dropped.
  await lane.event({ type: "failure", code: "backend-unavailable", recoverable: true, message: "late" })
  expect(a.frames.length).toBe(1)

  const unreaped = recorder()
  await new TerminalFrameLane(unreaped.sink, "c-2").event({ type: "exit", known: false, code: null, signal: null })
  expect(unreaped.frames[0]).toBe(SERVER_FIXTURES.exitUnknown)

  const signalled = recorder()
  await new TerminalFrameLane(signalled.sink, "c-3").event({ type: "exit", known: true, code: null, signal: 9 })
  expect(signalled.frames[0]).toBe(SERVER_FIXTURES.exitSignalled)

  const lost = recorder()
  const lostLane = new TerminalFrameLane(lost.sink, "c-4")
  await lostLane.event({
    type: "failure",
    code: "backend-unavailable",
    recoverable: true,
    message: "zmx helper exited before the attach completed",
  })
  expect(lost.frames).toEqual([SERVER_FIXTURES.failureRecoverableSayingExit])
  // Crucially: not an exit frame, however much the text talks about exiting.
  expect(lost.frames.some((f) => typeof f === "string" && JSON.parse(f).type === "exit")).toBe(false)
})

test("owner frames stamp the live epoch and a rising generation", async () => {
  const { frames, sink } = recorder()
  const lane = new TerminalFrameLane(sink, "c-1")
  await lane.ready()
  await lane.event({ type: "reset", epoch: "e-a" })
  await lane.event({ type: "replay-start", epoch: "e-a" })
  await lane.event({ type: "replay-end", epoch: "e-a" })
  await lane.event({ type: "owner", enabled: true })
  expect(frames.at(-1)).toBe(SERVER_FIXTURES.ownerOn)
  expect(lane.replyOwner).toBe(true)
  expect(lane.ownerGeneration).toBe(1)
  await lane.event({ type: "owner", enabled: false })
  expect(JSON.parse(String(frames.at(-1)))).toEqual({ type: "owner", epoch: "e-a", enabled: false, ownerGeneration: 2 })
})

test("a reply is accepted only past replay-end, while owning, on the live epoch and generation", async () => {
  const { sink } = recorder()
  const lane = new TerminalFrameLane(sink, "c-1")
  const reply = (epoch: string, generation: number) =>
    ({ type: "reply", epoch, ownerGeneration: generation, data: "G1syNDsxUg==" }) as const
  await lane.ready()
  await lane.event({ type: "reset", epoch: "e-a" })
  await lane.event({ type: "replay-start", epoch: "e-a" })
  // An answer produced while the replay is still being drawn is an answer to
  // HISTORY; the shell would read it as typing.
  await lane.event({ type: "owner", enabled: true })
  expect(lane.acceptsReply(reply("e-a", 1))).toBe(false)
  await lane.event({ type: "replay-end", epoch: "e-a" })
  expect(lane.acceptsReply(reply("e-a", 1))).toBe(true)
  // Stale epoch, stale generation, and a viewer that has lost the size.
  expect(lane.acceptsReply(reply("e-gone", 1))).toBe(false)
  expect(lane.acceptsReply(reply("e-a", 0))).toBe(false)
  await lane.event({ type: "owner", enabled: false })
  expect(lane.acceptsReply(reply("e-a", 2))).toBe(false)
})

test("viewer control frames decode by shape, not by substring", () => {
  expect(decodeClientControl(CLIENT_FIXTURES.resize)).toEqual({ ok: true, frame: { type: "resize", cols: 120, rows: 40 } })
  expect(decodeClientControl(CLIENT_FIXTURES.focus)).toEqual({ ok: true, frame: { type: "focus", focused: true, cols: 120, rows: 40 } })
  expect(decodeClientControl(CLIENT_FIXTURES.blur)).toEqual({ ok: true, frame: { type: "focus", focused: false, cols: 0, rows: 0 } })
  expect(decodeClientControl(CLIENT_FIXTURES.close)).toEqual({ ok: true, frame: { type: "close" } })
  const reply = decodeClientControl(CLIENT_FIXTURES.reply)
  expect(reply.ok).toBe(true)
  if (reply.ok && reply.frame.type === "reply") {
    expect(reply.frame.epoch).toBe("e-a")
    expect(reply.frame.ownerGeneration).toBe(1)
    expect(decodeReplyPayload(reply.frame.data)).toEqual(new TextEncoder().encode("\u001b[24;1R"))
  }
  // A string where a number belongs used to reach the pty as NaN.
  expect(decodeClientControl(CLIENT_FIXTURES.resizeNotANumber).ok).toBe(false)
  expect(decodeClientControl(SERVER_FIXTURES.malformed)).toEqual({ ok: false, reason: "malformed json" })
  expect(decodeClientControl(`[]`).ok).toBe(false)
  expect(decodeClientControl(SERVER_FIXTURES.unknownType).ok).toBe(false)
  expect(decodeReplyPayload("!!not base64!!")).toBe(null)
})

test("every fixture is decoded by the Kotlin client too", () => {
  const kotlin = readFileSync(
    join(import.meta.dir, "../apps/shared/src/commonTest/kotlin/dev/supermux/net/TerminalProtocolTest.kt"),
    "utf8",
  )
  const missing: string[] = []
  for (const [name, fixture] of [...Object.entries(SERVER_FIXTURES), ...Object.entries(CLIENT_FIXTURES)]) {
    // Kotlin holds these in raw ("""…""") strings, where the fixture appears
    // verbatim; an escaped literal would appear with `\"` for every quote.
    // Accept either spelling, as long as the BYTES are the same.
    const escaped = fixture.replace(/\\/g, "\\\\").replace(/"/g, '\\"')
    if (!kotlin.includes(fixture) && !kotlin.includes(escaped)) missing.push(`${name}: ${fixture}`)
  }
  expect(missing).toEqual([])
})
